src/main/java/org/assertj/core/internal/DeepDifference.java

```diff
@@ -443,6 +464,44 @@ private static boolean compareArrays(Object array1, Object array2, List<String>
   }
 
   /**
+   * Deeply compares two collections referenced by dualKey. This method attempts
+   * to quickly determine inequality by length, then if lengths match, in case of
+   * collection type is Set and there are passed no custom comparators, there is used
+   * comparison on hashcodes basis, otherwise each element from one collection is checked
+   * for existence in another one using 'deep' comparison.
+   */
+  private static <K, V> boolean compareUnorderedCollection(Collection<K> col1, Collection<V> col2,
+                                                           List<String> path, Deque<DualKey> toCompare,
+                                                           Set<DualKey> visited,
+                                                           Map<String, Comparator<?>> comparatorByPropertyOrField,
+                                                           TypeComparators comparatorByType) {
+    if (col1.size() != col2.size()) {
+      return false;
+    }
+
+    boolean hasNoCustomComparators = (comparatorByPropertyOrField == null || comparatorByPropertyOrField.isEmpty())
+      && (comparatorByType == null || comparatorByType.isEmpty());
+    if (hasNoCustomComparators && col1 instanceof Set) {
+      // this comparison is used for performance optimization reasons
+      return compareUnorderedCollectionByHashCodes(col1, col2, path, toCompare, visited);
+    }
+
+    Collection<V> col2Copy = new LinkedList<>(col2);
+    for (Object o1 : col1) {
+      Iterator<V> iterator = col2Copy.iterator();
+      while (iterator.hasNext()) {
+        Object o2 = iterator.next();
+        if (determineDifferences(o1, o2, path, comparatorByPropertyOrField, comparatorByType).isEmpty()) {
```

Comment:
probably here (and some other places) could have been added something to the path, to distinguish collection element from collection itself. (May be needed in case there are custom comparators by field names for collection elements)